adminbot.py

```diff
@@ -47,17 +56,33 @@ def call_help(self, op):
         # this is going to be handled by a listener plugin
         pass
 
+@auth.verify_password
+def verify_password(username, password):
+    if username in USERS:
+        return check_password_hash(USERS.get(username), password)
+    return False
+
+with open('pw.secret') as p:
+    pw = p.read().strip()
+with open('channels.json') as c:
+    channels = json.load(c)
+ops = [
+    'vilmibm',
+    'archangelic',
+    'equa',
+    'l0010o0001l'
+]
+bot = AdminBot(channels, 'adminbot', 'localhost', ops=ops, ns_pass=pw)
+
+@app.route('/notify', methods=['POST'])
+@auth.login_required
+def notify_admins():
+    channel = request.json['channel']
+    message = request.json['message']
+    bot.connection.privmsg(channel, message)
 
 if __name__ == '__main__':
-    with open('pw.secret') as p:
```

Comment:
out of curiosity, why did this code move to the top level?

Reply:
Programming is still 90% magic to me. I wasn’t certain if I needed to initialize the bot before doing the flask route which uses `bot`. So I moved it up to do that.

Reply:
ahhh ok, i see now.

some thoughts, and forgive me if i'm being splainy or overly pedantic:

you want an instance of `bot` accessible whenever Flask calls `notify_admins`. while the code you have here works since `notify_admins` _closes_ over the variable `bot`, it means that every time this file is interpreted by Python it's going to run that setup code. You might not want that behavior, especially if this file is ever intended to be imported and extended (i.e. instead of run from the command line, triggering that `if __name__ == '__main__'` code).

I would recommend:

- create a top level variable `bot` and initially set it to `None` at the top of the file
- write a top level function called `get_bot` that returns the variable `bot` or raises an error saying "bot isn't initialized, i cannot do anything for you"
- resume using your `if __name__ == '__main__'` setup block. when you set `bot`, you'll be setting the globally visible `bot` variable.